pkg/server/server_test.go

```diff
@@ -200,6 +201,80 @@ func TestCheckDoesNotThrowBecauseDirectTupleWasFound(t *testing.T) {
 	require.Equal(t, true, checkResponse.Allowed)
 }
 
+func TestListObjectsReleasesConnections(t *testing.T) {
+	testDatastore := storagefixtures.RunDatastoreTestContainer(t, "postgres")
+
+	uri := testDatastore.GetConnectionURI(true)
+	ds, err := postgres.New(uri, sqlcommon.NewConfig(
+		sqlcommon.WithMaxOpenConns(1),
+		sqlcommon.WithMaxTuplesPerWrite(2000),
+	))
+	require.NoError(t, err)
+	defer ds.Close()
+
+	s := MustNewServerWithOpts(
+		WithDatastore(storagewrappers.NewContextWrapper(ds)),
+		WithMaxConcurrentReadsForListObjects(1),
+	)
+
+	storeID := ulid.Make().String()
+
+	writeAuthzModelResp, err := s.WriteAuthorizationModel(context.Background(), &openfgav1.WriteAuthorizationModelRequest{
+		StoreId: storeID,
+		TypeDefinitions: parser.MustParse(`
+		type user
+
+		type document
+		  relations
+		    define editor: [user] as self
+		`),
+		SchemaVersion: typesystem.SchemaVersion1_1,
+	})
+	require.NoError(t, err)
+
+	modelID := writeAuthzModelResp.GetAuthorizationModelId()
+
+	numTuples := 2000
+	tuples := make([]*openfgav1.TupleKey, 0, numTuples)
+	for i := 0; i < numTuples; i++ {
+		tk := tuple.NewTupleKey(fmt.Sprintf("document:%d", i), "editor", "user:jon")
+
+		tuples = append(tuples, tk)
+	}
+
+	_, err = s.Write(context.Background(), &openfgav1.WriteRequest{
+		StoreId:              storeID,
+		AuthorizationModelId: modelID,
+		Writes: &openfgav1.TupleKeys{
+			TupleKeys: tuples,
+		},
+	})
+	require.NoError(t, err)
+
+	cancelCtx, cancel := context.WithCancel(context.Background())
+
+	_, err = s.ListObjects(cancelCtx, &openfgav1.ListObjectsRequest{
+		StoreId:              storeID,
+		AuthorizationModelId: modelID,
+		Type:                 "document",
+		Relation:             "editor",
+		User:                 "user:jon",
+	})
+	require.NoError(t, err)
+	cancel()
+
+	timeoutCtx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
+	defer cancel()
+
+	checkResp, err := s.Check(timeoutCtx, &openfgav1.CheckRequest{
+		StoreId:              storeID,
+		AuthorizationModelId: modelID,
+		TupleKey:             tuple.NewTupleKey("document:0", "editor", "user:jon"),
+	})
+	require.NoError(t, err)
```

Comment:
This Check fails without the code changes in ReverseExpand, because it timesout due to there not being an available connection in the database connection pool, since it is being hogged by the ListObjects call still.